File: js/036_ValidSudoku.js
/**
 * @param {character[][]} board
 * @return {boolean}
 */
var isValidSudoku = function(board) {
    var rowHash = [], colHash = [], gridHash = [];
    for(var m = 0; m < 9; ++m){
        rowHash[m] = [];
        colHash[m] = [];
        gridHash[m] = [];
        for(var n = 0; n < 9; ++n){
            rowHash[m].push(0);
            colHash[m].push(0);
            gridHash[m].push(0);
        }
    }
    
    for(var i = 0; i < board.length; ++i){
        for(var j = 0; j < board[i].length; ++j){
            if(board[i][j] != '.'){
                var index = parseInt(board[i][j]) - 1;
                var gridIndex = Math.floor(i/3)*3 + Math.floor(j/3);
                console.log("index:" + index + " grid:" + gridIndex);
                if(rowHash[i][index] || colHash[j][index] || gridHash[gridIndex][index]) return false;
                rowHash[i][index] = colHash[j][index] = gridHash[gridIndex][index] = 1;
            }
        }
    }
    return true;
};

File: cpp/036_ValidSudoku.cpp
class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        int rowHash[9][9] = {0}, colHash[9][9] = {0}, gridHash[9][9] = {0};
        for(int i = 0; i < board.size(); ++i){
            for(int j = 0; j < board[i].size(); ++j){
                if(board[i][j] != '.'){
                    int index = board[i][j]-'0'-1;
                    int gridIndex = i/3 * 3 + j/3;
                    if(rowHash[i][index] || colHash[j][index] || gridHash[gridIndex][index]) return false;
                    rowHash[i][index] = colHash[j][index] = gridHash[gridIndex][index] = 1;
                }
            }
        }
        return true;
    }
};